larq_compute_engine/core/indirect_bgemm/prepare.h

```diff
@@ -0,0 +1,145 @@
+
+#include <cstdint>
+
+#include "larq_compute_engine/core/types.h"
+#include "larq_compute_engine/tflite/kernels/bconv2d_params.h"
+#include "tensorflow/lite/c/builtin_op_data.h"
+#include "tensorflow/lite/kernels/internal/types.h"
+
+using namespace tflite;
+
+namespace compute_engine {
+namespace core {
+namespace indirect_bgemm {
+
+// This function is (heavily) adapted from this XNNPack function:
+// https://github.com/google/XNNPACK/blob/80a8ac59849bfdae8d2e1409f5642baa502c0b9e/src/indirection.c#L18-L76
+void FillIndirectionBuffer(const int block_size_pixels,
+                           const TfLiteBConv2DParams* conv_params,
+                           const RuntimeShape& bitpacked_input_shape,
+                           const RuntimeShape& output_shape,
+                           const TBitpacked* input_ptr,
+                           std::vector<const TBitpacked*>& indirection_buffer,
+                           std::vector<TBitpacked>& zero_buffer) {
+  using std::int32_t;
+
+  const int32_t kernel_height = conv_params->filter_height;
+  const int32_t kernel_width = conv_params->filter_width;
+  const int32_t stride_height = conv_params->stride_height;
+  const int32_t stride_width = conv_params->stride_width;
+  const int32_t dilation_height = conv_params->dilation_height_factor;
+  const int32_t dilation_width = conv_params->dilation_width_factor;
+  const int32_t input_padding_top = conv_params->padding_values.height;
+  const int32_t input_padding_left = conv_params->padding_values.width;
+
+  const int32_t input_height = bitpacked_input_shape.Dims(1);
+  const int32_t input_width = bitpacked_input_shape.Dims(2);
+  const int32_t bitpacked_input_channels = bitpacked_input_shape.Dims(3);
+
+  const int32_t output_height = output_shape.Dims(1);
+  const int32_t output_width = output_shape.Dims(2);
+
+  const int32_t output_size = output_height * output_width;
+  const int32_t kernel_size = kernel_height * kernel_width;
+  const int32_t tiled_output_size =
+      block_size_pixels *
+      ((output_size + block_size_pixels - 1) / block_size_pixels);
+
+  indirection_buffer.resize(tiled_output_size * kernel_size);
+  zero_buffer.assign(kernel_size * bitpacked_input_channels, 0);
+
+  for (int32_t output_tile_start = 0; output_tile_start < tiled_output_size;
+       output_tile_start += block_size_pixels) {
+    for (int32_t output_tile_offset = 0; output_tile_offset < block_size_pixels;
+         output_tile_offset++) {
+      const int32_t output_index =
+          std::min(output_tile_start + output_tile_offset, output_size - 1);
+      const int32_t output_x = output_index % output_width;
+      const int32_t output_y = output_index / output_width;
+      for (int32_t kernel_y = 0; kernel_y < kernel_height; kernel_y++) {
+        const int32_t input_y = output_y * stride_height +
+                                kernel_y * dilation_height - input_padding_top;
+        if (0 <= input_y && input_y < input_height) {
```

Comment:
Not sure how performance critical this part is, so feel free to ignore this comment. But this should remove one comparison in the generated code:
```suggestion
        // (0 <= input_y && input_y < input_height)
        if ((unsigned)input_y < (unsigned)input_height) {
```

Reply:
Wow, I've not seen that trick before. How does it work exactly?

The functions in `prepare.h` aren't performance critical because they only run once at the start, so I think I might leave it as-is because it's easier to understand.

Reply:
I got this from https://github.com/python/cpython/pull/9784 which links to section 14.2 "Bounds Checking" in https://www.agner.org/optimize/optimizing_cpp.pdf for an in-depth explanation.

Reply:
Ah nice, that's cool, thanks for the links :)